### ISSUE TITLE
NUM_PLACEHOLDERS MUST NOT be zero

### ISSUE THREAD

**Issue by kazuho** [quicwg/base-drafts#2753]
Currently, the text says:
```
The default value is 0. However, this value SHOULD be set to a non-zero value by servers.
```

I think we should change the default to unlimited or to some sensible number, and also introduce a minimum permitted number.

Prioritization is not something a client can enforce, therefore servers that prefer to ignore or mishandle placeholders can just ignore / mishandle them, as well as other information for prioritization (e.g. weights and dependencies).

To rephrase, communicating the capability of handling just one of the features of prioritization has marginal benefit compared to the complexity it introduces to the client side (e.g., the need to use different prioritization approach depending on if the server supports placeholders).

Or, if it is the case that we do not expect browsers to no longer use placeholders but are keeping it for proxies, we should remove "SHOULD", because that is a minor use case that has not seen actual use in HTTP/2.

**Comment by dtikhonov** [comment]
The recommended value was 16 a few draft versions ago.  There should be some GitHub artefact justifying changing it to 0.

**Comment by MikeBishop** [comment]
Settings can only be increased by the SETTINGS frame, if the peer's behavior is bounded by (and not just informed by) the value.  The default therefore needs to be the lowest possible value.  I presume that it was changed when we created default setting values.

**Comment by MikeBishop** [comment]
Yes, that line last changed in #2038.

**Comment by kazuho** [comment]
Maybe the alternative approach would be to:
* Remove the settings parameter, or make it a purely advisory variable like MAX_HEADER_LIST_SIZE allowing the client to exceed the advertised maximum.
* Recommend some sensible number of placeholders that a server is expected to retain (16 - the default value we had prior to merging #2038 - might make sense).

These changes would essentially bring back what we observe in HTTP/2 - server that ignore placeholders would just ignore.

**Comment by dtikhonov** [comment]
> These changes would essentially bring back what we observe in HTTP/2 - server that ignore placeholders would just ignore.

Introduction of placeholders is the backward-compatible part of HTTP/3 prioritization changes: if the client does not use placeholders it can use HTTP/3 priorities without placeholders just as in HTTP/2.  In other words, the clients can ignore placeholders as well.

I agree that it seems that a server SHOULD provide some sensible value.  My guess -- based on some of my mental models -- is that this number should be between N/2 and N, where N is the maximum number of parallel requests the server allows.

**Comment by kazuho** [comment]
> I agree that it seems that a server SHOULD provide some sensible value. My guess -- based on some of my mental models -- is that this number should be between N/2 and N, where N is the maximum number of parallel requests the server allows.

FWIW the maximum we have been imposing is 16, we've been closing the connections that goes above that with an ENHANCE_YOUR_CALM error. I do not think that has caused interoperability issues.

**Comment by ianswett** [comment]
Currently, only Firefox would be required to change their prioritization scheme if the number of placeholders is 0.  Based on information I've seen(ie: the Cloudflare presentation and my own anecdotal experience), it's not clear their use of placeholders is any better than a scheme that does not use them, so I think it's possible Firefox could change their priority scheme and it would be faster.

Given that, and that placeholders are a new feature in H3, I'd prefer to allow the number to be 0.

**Comment by kazuho** [comment]
Duplicate of #2734?

**Comment by kazuho** [comment]
@ianswett 
> Currently, only Firefox would be required to change their prioritization scheme if the number of placeholders is 0. Based on information I've seen(ie: the Cloudflare presentation and my own anecdotal experience), it's not clear their use of placeholders is any better than a scheme that does not use them, so I think it's possible Firefox could change their priority scheme and it would be faster.

Fair point. I agree that placeholders becomes no longer a prerequisite for browsers once we bring back the "exclusive" bit (as proposed by #2754).

OTOH, I am still not sure if there is a necessity to diverge from the design we have in HTTP/2 and start requiring clients to take the server's settings into consideration when building a prioritization tree. We currently recommend servers to set `SETTINGS_NUM_PLACEHOLDERS` to some sensible value other than zero, and given that, as I commented in https://github.com/quicwg/base-drafts/issues/2753#issuecomment-495487631, suggesting server's to ignore placeholder IDs out of the server's advertised range might be a simpler solution.

**Comment by rmarx** [comment]
> Fair point. I agree that placeholders becomes no longer a prerequisite for browsers once we bring back the "exclusive" bit (as proposed by #2754).

@kazuho could you clarify this comment a bit? Do you mean placeholders are no longer needed if we bring back the exclusive bit **AND allow streams to depend on other streams**? I don't follow you if that last part isn't implied.

**Comment by kazuho** [comment]
@rmarx 
> Do you mean placeholders are no longer needed if we bring back the exclusive bit __AND allow streams to depend on other streams__?

Correct. IIUC, the context is if the browsers need to use a placeholder under current scheme (that already allows a stream to depend on another) with the exclusive bit being revived.

Let me be precise.

I think there is a rough consensus that a ideal tree looks like:

```
root
  +- JS1 (blocking)
      +- CSS1 (blocking)
        +- CSS2 (blocking)
            +- image1 (weight=X)
            +- image2 (weight=X)
            +- image3 (weight=X)
            +- image4 (weight=X)
```

HTML could be either within the root and JS1, or within CSS2 and image1.

The other thing that is missing in the chart above are the the asynchronously loaded scripts and CSS files. However, we could argue that prioritization of them is not important, as long as it receives comparable or less bandwidth than the images.

If we agree on that, there is no "need" for browsers to use placeholders (or idle streams as placeholders), neither in the HTTP/2 prioritization tree nor the current HTTP/3 scheme with the exclusive bit being revived.

**Comment by MikeBishop** [comment]
> **AND allow streams to depend on other streams**? I don't follow you if that last part isn't implied.

Streams are allowed to depend on other streams.  While @ianswett has proposed removing that in some of the do-better-than-H2 discussion, it hasn't been removed.

**Comment by ianswett** [comment]
I'm really happy that we have a fairly clear idea of what the goal is in terms of prioritization.  There's certainly a bit of wiggle room and room for experimentation, but I do believe we have rough consensus that we want a mechanism that can accomplish what's in @kazuho post.

As @kazuho says, if that's true, then I don't think we need placeholders to accomplish that prioritization.  And if that's true, then there's no critical dependency on a non-zero number of placeholders.
